packages/common/src/api/tan-query/events/useEvents.ts

```diff
@@ -0,0 +1,51 @@
+import { useMemo } from 'react'
+
+import { useQueryClient } from '@tanstack/react-query'
+import { keyBy } from 'lodash'
+import { useDispatch } from 'react-redux'
+
+import { useAudiusQueryContext } from '~/audius-query'
+import { Event } from '~/models/Event'
+import { ID } from '~/models/Identifiers'
+
+import { getEventsBatcher } from '../batchers/getEventsBatcher'
+import { QueryOptions } from '../types'
+import { useCurrentUserId } from '../useCurrentUserId'
+import { combineQueryResults } from '../utils/combineQueryResults'
+import { useQueries } from '../utils/useQueries'
+
+import { getEventQueryKey } from './utils'
+
+export const useEvents = (
+  eventIds: ID[] | null | undefined,
+  options?: QueryOptions
+) => {
+  const { audiusSdk } = useAudiusQueryContext()
+  const dispatch = useDispatch()
+  const queryClient = useQueryClient()
+  const { data: currentUserId } = useCurrentUserId()
+
+  const queryResults = useQueries({
+    queries: eventIds?.map((eventId) => ({
+      queryKey: getEventQueryKey(eventId),
+      queryFn: async () => {
+        const sdk = await audiusSdk()
+        const batchGetEvents = getEventsBatcher({
+          sdk,
+          currentUserId,
+          queryClient,
+          dispatch
+        })
+        return await batchGetEvents.fetch(eventId)
+      },
+      ...options,
+      enabled: options?.enabled !== false && !!eventId
+    })),
+    combine: combineQueryResults<Event[]>
+  })
+  const { data: events } = queryResults
+
+  const byId = useMemo(() => keyBy(events, 'eventId'), [events])
+
+  return { ...queryResults, byId }
```

Comment:
tan-query actually monitors which fields you access from the results, so it's best to avoid spreading the results object. If you do, it'll cause rerenders on any change to any part of the queryResults. If you don't, it'll only rerender when a field you access changes.

Here's a helper I've been meaning to add, but you can also just do it in line here. Just cast the queryResults and assign `queryResults.byId` directly.

```
export const addByIdQueryResults = <T>(
  queryResults: UseQueryResult<T[]>,
  key: keyof T
) => {
  const byId = keyBy(queryResults, key) as Record<ID, T>
  const queryResultsWithById = queryResults as UseQueryResult<T[]> & {
    byId: Record<ID, T>
  }
  queryResultsWithById.byId = byId
  return queryResultsWithById
}
```